File: leetcode/unique_paths.js
// https://leetcode.com/problems/unique-paths/
// 62. Unique Paths
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
let map = {};
function recurse(m,n) {
    if (n == 1) {
        return 1
    }
    let count = 1;
    for (let i = 2; i <= m; i++) {
        let sum = 0;
        if (map[`${Math.max(i,n-1)}${Math.min(i,n-1)}`]) {
            sum = map[`${Math.max(i,n-1)}${Math.min(i,n-1)}`]
        } else {
            sum = recurse(i,n-1);
            map[`${Math.max(i,n-1)}${Math.min(i,n-1)}`] = sum;
        }
        count += sum
    }
    return count;
}
var uniquePaths = function(m, n) {
    return recurse(m,n);

};

function main() {
    let res = uniquePaths(10,10);
    console.log(res);
}

main()

File: leetcode/unique_paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        grid = [[0] * n] * m
        grid[0][0] = 1
        def getCell(i,j):
            if i < 0 or i >= m or j < 0 or j >= n:
                return 0
            return grid[i][j]
        for i in range(m):
            for j in range(n):
                grid[i][j] = max(grid[i][j],getCell(i-1,j) + getCell(i,j-1))
        return grid[m-1][n-1]

if __name__ == '__main__':
    m = 3
    n = 7
    res = Solution().uniquePaths(m,n)
    print(res)